enhancements/20231108-elemental-cloudinit-crd.md

```diff
@@ -0,0 +1,176 @@
+# Kubernetes-native Elemental cloud-init documents
+
+## Summary
+
+It is both an expected and supported behavior for Harvester users to
+persist certain configurations to their Harvester nodes. These are currently
+done with Elemental cloud-init documents under the `/oem` path on the system.
+
+Currently, Harvester users would have to write these cloud-init documents
+and manually push them to each node where they would like to effect
+the desired change.
+
+There is a cloud-native tooling gap with regards to this process, however.
+Harvester users might prefer to use a GitOps-style approach to configuring
+their Harvester clusters, where Kubernetes objects are version-controlled
+as YAML and each new revision is applied to the cluster.
+
+This HEP presents a new Custom Resource Definition (CRD) that is surfaced
+via the harvester-node-manager, where a Harvester user could apply a "CloudInit"
+CR to a Harvester cluster, like so:
+
+```yaml
+apiVersion: node.harvesterhci.io/v1beta1
+kind: CloudInit
+metadata:
+  name: ssh-access
+spec:
+  matchSelector: {}
+  filename: 99_ssh.yaml
+  contents: |
+    stages:
+      network:
+        - authorized_keys:
+            rancher:
+              - ssh-ed25519 AAAA...
+```
+
+### Related Issues
+
+https://github.com/harvester/harvester/issues/4712
+
+## Motivation
+
+Provide a GitOps-compatible way of persisting configuration changes to
+a Harvester cluster.
+
+### Goals
+
+* Elemental cloud-init documents are stored in a CloudInit object spec and
+  pulled down onto nodes that match the CloudInit's match selector.
+* CloudInit object's status field describes how it is rolled out
+  across the cluster (for example: which nodes does it apply to,
+  are any out of sync, etc).
+
+### Non-goals [optional]
+
+* Automatic rollbacks. If a Harvester user wants to "undo" a previously applied
+  CloudInit object, they should expect to:
+
+  1. Revise the CloudInit object contents and reboot; or
+  2. Adjust the CloudInit object match selector to exclude a node that they
+     no longer want it to apply to and remove the file under `/oem` and reboot; or
+  3. Delete the CloudInit object and then remove the underlying file from `/oem`
+     on affected nodes and reboot the nodes at their leisure.
+
+* Automatic execution. Harvester users should expect to reboot a
+  Harvester node to allow the resulting cloud-init documents to be
+  applied at boot.
+
+## Proposal
+
+### User Stories
```

Comment:
events may not be the best place to record this info, since the default TTL is 1hour.

Reply:
We may keep both POD log and event.

For production depoyment, many customers have third-party logging server, which filters&stores logging/event from Harvester cluster, then store them for longer time.

Reply:
Is it sufficient for this information to surface as Conditions?

```
Status:
  Rollouts:
    Conditions:
      Last Transition Time:  2023-10-24T21:15:02Z
      Message:               
      Reason:                CloudInitApplicable
      Status:                True
      Type:                  Applicable
      Last Transition Time:  2023-10-24T21:09:28Z
      Message:               Local file checksum is the same as the CloudInit checksum
      Reason:                CloudInitChecksumMatch
      Status:                False
      Type:                  OutOfSync
      Last Transition Time:  2023-10-24T21:09:28Z
      Message:               99_ssh.yaml is present under /oem
      Reason:                CloudInitPresentOnDisk
      Status:                True
      Type:                  Present
    Node Name:               harv1
```

Reply:
When user update the same CRD object many times, the Conditions normally only reflect the last (few) changes.

Reply:
Thanks! I've updated the user stories with some examples. I could not be exhaustive about everything a user could possibly do to a Harvester node since they are only limited by their imagination and the cloud-init syntax.

I've also added a note indicating that the controller will create a kubernetes event whenever it modifies a file under `/oem` and it will also indicate which node it happened on.